keps/sig-apps/592-ttl-after-finish/README.md

```diff
@@ -250,17 +257,141 @@ Mitigations:
-with no changes. 
 
 [umbrella issues]: https://github.com/kubernetes/kubernetes/issues/42752
 
-## Implementation History
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+* **How can this feature be enabled / disabled in a live cluster?**
+  - [x] Feature gate (also fill in values in `kep.yaml`)
+    - Feature gate name: TTLAfterFinished
+    - Components depending on the feature gate: kube-apiserver, kube-controller-manager
+  - [ ] Other
+    - Describe the mechanism:
+    - Will enabling / disabling the feature require downtime of the control
+      plane?
+    - Will enabling / disabling the feature require downtime or reprovisioning
+      of a node? (Do not assume `Dynamic Kubelet Config` feature is enabled).
+
+* **Does enabling the feature change any default behavior?**
+  No.
+
+* **Can the feature be disabled once it has been enabled (i.e. can we roll back
+  the enablement)?**
+  Yes.
+
+* **What happens if we reenable the feature if it was previously rolled back?**
+  It should work as expected.
+
+* **Are there any tests for feature enablement/disablement?**
+  No.
+
+### Rollout, Upgrade and Rollback Planning
+
+* **How can a rollout fail? Can it impact already running workloads?**
+  It shouldn't impact already running workloads. This is an opt-in feature since users need to 
+  explicitly set the TTLSecondsAfterFinished parameter in the job spec, which is accepted by the 
+  api-server only if the feature is enabled.
+
+* **What specific metrics should inform a rollback?**
+  Unexpected restarts of kube-controller-manager
+
+* **Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?**
+  N/A
+
+* **Is the rollout accompanied by any deprecations and/or removals of features, APIs, 
+fields of API types, flags, etc.?**
+  No
+
+### Monitoring Requirements
+
+_This section must be completed when targeting beta graduation to a release._
+
+* **How can an operator determine if the feature is in use by workloads?**
+  - Check the workqueue metrics
+  - When checking the job object in the api server, the TTLSecondsAfterFinished parameter should 
+    be set if it was specified when the job was created.
+
+* **What are the SLIs (Service Level Indicators) an operator can use to determine 
+the health of the service?**
+ - [x] Metrics
+    - Metric name: `ttl_after_finished_controller_rate_limiter_use`
+    - Metric name: `workqueue_adds_total`
+    - Metric name: `workqueue_depth`
+    - Metric name: `workqueue_queue_duration_seconds`
+    - Metric name: `workqueue_retries_total`
+    - Components exposing the metric: `kube-controller-manager`
+    - Metric name: `etcd_object_counts{resource="jobs.batch"}`
+    - Components exposing the metric: `kube-apiserver`.
+
+* **What are the reasonable SLOs (Service Level Objectives) for the above SLIs?**
+
+99% of the jobs that needs cleanup are deleted within X minutes.
```

Comment:
I don't see an easy way to compute such SLO out of the existing metrics.

Can we create a metric that will basically expose this information?
Something like:
"histogram of when the job were supposed to be deleted minus when it was deleted"
[somewhat similar to what we did (or supposed to do) for cronjob: https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/19-Graduate-CronJob-to-Stable#metrics ]

Reply:
I was thinking we could use `workqueue_queue_duration_seconds`: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/metrics/prometheus/workqueue/metrics.go#L60

Reply:
Can you describe how? [I think you will be putting all updates to Job objects in the queue, so distinguishing the ones that are relevant might be hard in that metric.]

Reply:
Thanks for explanation.

That said - I don't think you can use workqueue metrics to report what we want exactly:
1) We really want to wait for TTL - we don't want to delete jobs too early
2) The fact that the queue is keeping up with load doesn't mean there isn't a bug in the code and we aren't really deleting as fast as we should.

So I really think we should add a metric similar to what was done for CronJob...

Reply:
> We really want to wait for TTL - we don't want to delete jobs too early

We are not deleting jobs too early. Not sure how you concluded that from my comment.

> The fact that the queue is keeping up with load doesn't mean there isn't a bug in the code and we aren't really deleting as fast as we should.

Where would the metric be updated? To achieve the semantics you are looking for I think we need a dedicated metric endpoint such that when queried, all jobs are listed and a histogram is computed (something like https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/1748-pod-resource-metrics).

Reply:
> We are not deleting jobs too early. Not sure how you concluded that from my comment.

Sorry - I didn't mean we're deleting.
What I was saying: "what if there is a bug that leads to deleting them to early".

In general, the purpose of many of metrics isn't needed is everything works as designed. They are used if there are bugs, something doens't work as expected, etc....

> Where would the metric be updated? To achieve the semantics you are looking for I think we need a dedicated metric endpoint such that when queried, all jobs are listed and a histogram is computed (something like https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/1748-pod-resource-metrics).

What I had on mind is basically:
- when you send a "DELETE job" request, you basically export the metric  "currentTime - time when job succeeded - TTL value"

Sure - if the feature doesn't work at all we won't see anything in metrics. But assuming it eventually deletes stuff, we will see how accurately it does.

Reply:
> In general, the purpose of many of metrics isn't needed is everything works as designed. They are used if there are bugs, something doens't work as expected, etc....

Right, but my general argument is that adding a new metric isn't going to make things a lot different compared to relying on the workqueue ones with respect to avoiding bugs.

 > when you send a "DELETE job" request, you basically export the metric "currentTime - time when job succeeded - TTL value"
> Sure - if the feature doesn't work at all we won't see anything in metrics. But assuming it eventually deletes stuff, we will see how accurately it does.

This doesn't protect us from bugs in the workqueue (your initial concern about the workqueue metrics), what if the workqueue has a bug were push sometimes work and sometimes silently doesn't (and so some jobs will get deleted and some will not or randomly delayed)? 

In any case, I updated the KEP to indicate that we will add this metric. But, our source of truth is etcd, and I think the only way to reliably achieve the semantics you want is by querying that directly.

Reply:
I agree that doesn't protect us from all kind of bugs, but I think that metric itself is fairly useful on its own.

Anyway - thanks!